File: implementations/java-springboot/harness/src/rules/DomainLayerIsolation.java
package harness.rules;

import harness.Finding;
import harness.RuleResult;

import java.io.File;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import static harness.JavaFiles.collectJavaFiles;
import static harness.JavaFiles.pathContains;
import static harness.JavaFiles.readText;
import static harness.JavaFiles.relTo;

/**
 * [13] Domain layer isolation — a file inside {@code <domain>/domain/} may not import its
 * own {@code application/}/{@code infrastructure/}/{@code interfaces/}, nor the same
 * layers of a sibling domain (layer-architecture.md — "upper layers may depend on lower
 * layers, but not the reverse").
 *
 * <p>This is a path (package)-based check — unlike the {@code domain-purity} rule, which
 * only blocklists specific Spring-annotation strings (`@Service`/`@Component`/...), this
 * rule hardcodes no framework name at all and looks only at "does the import statement's
 * package path contain an application/infrastructure/interfaces segment" — so it also
 * catches domain referencing an arbitrary upper-layer class that isn't Spring-related
 * (a more structural/comprehensive check).
 */
public final class DomainLayerIsolation {
    private DomainLayerIsolation() {
    }

    private static final Pattern IMPORT_LINE = Pattern.compile("^import\\s+(?:static\\s+)?([\\w.]+);", Pattern.MULTILINE);
    private static final Pattern FORBIDDEN_SEGMENT = Pattern.compile("\\.(application|infrastructure|interfaces)\\.");

    public static RuleResult check(String rootPath) {
        File root = new File(rootPath);
        RuleResult result = new RuleResult("domain-layer-isolation");
        boolean found = false;

        for (File f : collectJavaFiles(root)) {
            if (!pathContains(f, "/domain/")) continue;
            found = true;
            String rel = relTo(f, root);
            String code = readText(f);

            Matcher importMatcher = IMPORT_LINE.matcher(code);
            String violation = null;
            while (importMatcher.find()) {
                String imported = importMatcher.group(1);
                if (FORBIDDEN_SEGMENT.matcher(imported).find()) {
                    violation = imported;
                    break;
                }
            }

            if (violation != null) {
                result.add(Finding.fail(rel,
                    "A domain/ class imports an upper layer — '" + violation
                        + "' (application/infrastructure/interfaces may reference domain, never the reverse, layer-architecture.md)"));
            } else {
                result.add(Finding.pass(rel + " (confirmed domain isolation)"));
            }
        }

        if (!found) result.add(Finding.skip("No Java files under domain/"));
        return result;
    }
}

File: implementations/kotlin-springboot/harness/src/rules/DomainLayerIsolation.kt
package harness.rules

import harness.*
import java.io.File

// Catching a mention inside a comment would be a false positive, so line/block comments are stripped
// first to look only at real import declarations.
private val BLOCK_COMMENT = Regex("""/\*[\s\S]*?\*/""")
private val LINE_COMMENT = Regex("""//[^\n]*""")

private fun stripComments(content: String): String =
    content.replace(BLOCK_COMMENT, "").replace(LINE_COMMENT, "")

// A structural(path-based) check that doesn't hardcode domain names — an import of the form
// com.example.accountservice.<any BC>.(application|infrastructure|interfaces) is forbidden regardless
// of whether that BC is domain/'s own BC or a different(sibling) one. Broader in scope than the
// domain-purity rule(a blocklist of framework annotations/JPA imports) — this enforces the dependency
// direction itself, that domain/ must know about no higher-layer code at all.
private val FORBIDDEN_LAYER_IMPORT =
    Regex("""^import\s+com\.example\.accountservice\.\w+\.(application|infrastructure|interfaces)\b""", RegexOption.MULTILINE)

/**
 * [R1] domain-layer-isolation — a domain/ file may not import the application/, infrastructure/,
 * interfaces/ package(whether its own BC or a sibling BC) (layer-architecture.md).
 */
fun checkDomainLayerIsolation(rootPath: String): RuleResult {
    val root = File(rootPath)
    val result = RuleResult("domain-layer-isolation")
    var found = false

    for (f in collectKtFiles(root)) {
        if (!f.pathContains("/domain/")) continue
        found = true
        val rel = f.relTo(root)
        val code = stripComments(f.readText())
        val violation = FORBIDDEN_LAYER_IMPORT.find(code)
        if (violation != null) {
            result.add(
                failFinding(
                    rel,
                    "domain/ may not import application/·infrastructure/·interfaces/(including its own or another domain) — '${violation.value.trim()}' (layer-architecture.md)",
                ),
            )
        } else {
            result.add(passFinding("$rel (domain layer isolated)"))
        }
    }

    if (!found) result.add(skipFinding("no domain/ Kotlin files"))
    return result
}
